resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -229,8 +229,17 @@ private[spark] class ApplicationMaster(
         }
 
         if (!unregistered) {
+          // Only when ApplicationMaster's exit code is "EXIT_REPORTER_FAILURE" or "EXIT_EARLY" or
+          // "EXIT_UNCAUGHT_EXCEPTION" we should let RM to do reattempt. If the exception is from
+          // Spark internally, for example illegal state, executor failure, user code bug, we
+          // should not trigger another attempt.
+          val shouldUnregister = !(exitCode == ApplicationMaster.EXIT_REPORTER_FAILURE ||
+            exitCode == ApplicationMaster.EXIT_EARLY ||
+            exitCode == ApplicationMaster.EXIT_UNCAUGHT_EXCEPTION)
```

Comment:
Move this into a separate method ?
Also, can we make it explicit in what the exitCode handled is `exitCode == EXIT_EXCEPTION_USER_CLASS || exitCode == v2 ..` instead of negation ?